src/scheduled_tasks.rs

```diff
@@ -356,48 +357,156 @@ async fn overwrite_deleted_posts_and_comments(pool: &mut DbPool<'_>) {
-        ("1 day", "day"),
-        ("1 week", "week"),
-        ("1 month", "month"),
-        ("6 months", "half_year"),
-      ];
-
-      for i in &intervals {
-        let update_site_stmt = format!(
-      "update site_aggregates set users_active_{} = (select * from r.site_aggregates_activity('{}')) where site_id = 1",
-      i.1, i.0
-    );
-        sql_query(update_site_stmt)
-          .execute(&mut conn)
-          .await
-          .map_err(|e| error!("Failed to update site stats: {e}"))
-          .ok();
-
-        let update_community_stmt = format!("update community_aggregates ca set users_active_{} = mv.count_ from r.community_aggregates_activity('{}') mv where ca.community_id = mv.community_id_", i.1, i.0);
-        sql_query(update_community_stmt)
-          .execute(&mut conn)
-          .await
-          .map_err(|e| error!("Failed to update community stats: {e}"))
-          .ok();
-      }
-
-      info!("Done.");
+      process_site_aggregates(&mut conn, interval).await;
+      process_community_aggregates(&mut conn, interval).await;
     }
     Err(e) => {
       error!("Failed to get connection from pool: {e}");
     }
   }
 }
 
+/// Re-calculate all the active counts
+async fn all_active_counts(pool: &mut DbPool<'_>, intervals: [(&str, &str); 4]) {
+  for i in intervals {
+    active_counts(pool, i).await;
+  }
+}
+
+async fn process_site_aggregates(conn: &mut AsyncPgConnection, interval: (&str, &str)) {
+  // Select the site count result first
+  let site_activity_res = sql_query(format!(
+    "select * from r.site_aggregates_activity('{}')",
+    interval.0
+  ))
+  .get_result::<SiteActivitySelectResult>(conn)
+  .await;
+
+  match site_activity_res {
+    Ok(res) => {
+      let processed_rows = res.site_aggregates_activity;
+
+      // Update the site count
+      sql_query(format!(
+        "update site_aggregates set users_active_{} = $1 where site_id = 1",
+        interval.1,
+      ))
+      .bind::<Integer, _>(processed_rows)
+      .execute(conn)
+      .await
+      .map_err(|e| error!("Failed to update site stats: {e}"))
+      .ok();
+
+      info!(
+        "Finished site_aggregates active_{} (processed {} rows)",
+        interval.1, processed_rows
+      );
+    }
+    Err(e) => {
+      error!("Failed to fetch site activity: {e}");
+    }
+  }
+}
+
+async fn process_community_aggregates(conn: &mut AsyncPgConnection, interval: (&str, &str)) {
+  // Select the community count results into a temp table.
+  let caggs_temp_table = "community_aggregates_temp_table";
+
+  let create_table_res = sql_query(format!(
+    "CREATE TEMP TABLE {caggs_temp_table} AS SELECT * FROM r.community_aggregates_activity('{}')",
+    interval.0
+  ))
+  .execute(conn)
+  .await;
+
+  match create_table_res {
+    Ok(_) => {
+      // Split up into 1000 community transaction batches
+      let update_batch_size = 1000;
+      let mut processed_rows_count = 0;
+      let mut prev_community_id_res = Some(0);
```

Comment:
Did you measure how long it takes to update all communities in a single query? As it doesnt need to calculate anything in this case, it might be fast enough. Or possibly use a larger batch size.

Reply:
I just did it that way before this most revent commit, but with that method, its impossible to use a single `update from` query; it needed to do many individual update statements. Using a temp table is much cleaner.

> Did you measure how long it takes to update all communities in a single query? As it doesnt need to calculate anything in this case, it might be fast enough. Or possibly use a larger batch size.

You're correct that it might not even be necessary... but it can't hurt anything either. I'll test with lemmy.ml prod data sometime soon.